### PR TITLE
Fix CodeLlama tokenizer stripping real leading spaces on decode (#46491)

### LINKED ISSUE #46491: CodeLlama tokenizer strips one leading space on encode→decode round-trip (regression vs v4)
### System Info

- `transformers` version: 5.10.2
- `tokenizers` version: 0.22.2
- Python version: 3.11.2
- Platform: Linux (Penn State HPC, RHEL 8.10)
- PyTorch version: not relevant (tokenizer-only repro)

### Who can help?

@ArthurZucker @itazap

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

# CodeLlama tokenizer drops one leading space on decode round-trip (v5 regression)
 
## Reproduction
 
```python
from transformers import AutoTokenizer
 
tok = AutoTokenizer.from_pretrained("codellama/CodeLlama-7b-hf")
for s in ["   leading spaces", "    indented_line"]:
    decoded = tok.decode(tok.encode(s, add_special_tokens=False))
    print(f"{'OK' if decoded == s else 'BUG'}: {s!r} -> {decoded!r}")
```
 
Output on `transformers 5.10.2` + `tokenizers 0.22.2`:
 
```
BUG: '   leading spaces' -> '  leading spaces'
BUG: '    indented_line' -> '   indented_line'
```
 
The same script on `transformers 4.55.4` (last v4) round-trips correctly:
 
```
OK: '   leading spaces' -> '   leading spaces'
OK: '    indented_line' -> '    indented_line'
```
 
Exactly one leading space is lost on round-trip; trailing whitespace, internal whitespace, and other characters round-trip correctly.
 
## Scope
 
All CodeLlama variants tested are affected:
 
| Checkpoint | Class | Round-trips |
|---|---|---|
| `codellama/CodeLlama-7b-hf` | `CodeLlamaTokenizer` | 1/4 |
| `codellama/CodeLlama-7b-Instruct-hf` | `CodeLlamaTokenizer` | 1/4 |
| `codellama/CodeLlama-7b-Python-hf` | `CodeLlamaTokenizer` | 1/4 |
| `codellama/CodeLlama-13b-hf` | `CodeLlamaTokenizer` | 1/4 |
| `codellama/CodeLlama-34b-hf` | `CodeLlamaTokenizer` | 1/4 |
| `codellama/CodeLlama-70b-hf` | `LlamaTokenizer` | 1/4 |
 
## Root cause
 
`src/transformers/models/code_llama/tokenization_code_llama.py`, lines 163–164:
 
```python
self._tokenizer.decoder = decoders.Sequence(
    [decoders.Replace("▁", " "), decoders.ByteFallback(), decoders.Fuse(), decoders.Strip(content=" ", left=1)]
)
```
 
The trailing `decoders.Strip(content=" ", left=1)` unconditionally strips one leading space on decode. This was intended to remove the synthetic SentencePiece `▁` prefix introduced by the normalizer (`Prepend ▁` → `Replace " " → ▁`). But it also eats one leading space when the user's input itself starts with a space, because by that point in the decode pipeline the synthetic prefix and user-provided leading spaces are indistinguishable.
 
v4's `CodeLlamaTokenizerFast` had different decode handling that distinguished these. The v5 migration that removed the Fast class (#40936 "rm slow tokenizers", #42563 "Refactor-tokenization-more", #42894 "use TokenizersBackend") landed this Strip-step decoder for all CodeLlama checkpoints.
 
## Impact
 
This silently corrupts whitespace-sensitive workloads on CodeLlama. In code generation specifically, where indentation is semantically meaningful:
 
```python
input  = "    if x > 0:\n        return x"
output = tok.decode(tok.encode(input, add_special_tokens=False))
# output = "   if x > 0:\n        return x"   # 3 spaces instead of 4 — silent indentation corruption
```
 
Any pipeline using `encode → decode` for round-trip text comparison (BLEU/chrF/exact-match metrics, RL reward signals on code tokens, fine-tuning data validation) produces silently-wrong values when leading whitespace is present.
 
## Suggested fix direction
 
Two possible shapes; deferring to maintainers:
 
1. **Track the synthetic prefix.** Instead of `Strip(left=1)`, the decoder should only strip a leading space when the corresponding encoded sequence began with the synthetic `▁` prefix and *not* with a user-provided leading space. This requires the decoder to know whether `add_prefix_space` would have fired.
2. **Replace with v4-style decode logic.** Port whatever `CodeLlamaTokenizerFast.decode` did in v4 (which round-tripped correctly) into the current implementation.
Happy to implement once a direction is preferred.

### Expected behavior

`tok.decode(tok.encode(s))` should round-trip leading whitespace, as it does on transformers v4 and on most other tokenizers in v5 (Qwen, StarCoder2, Phi-3 all round-trip correctly on the same test strings).
 
This is a tokenization-correctness bug. Code generation and any whitespace-sensitive evaluation pipeline silently produce wrong values on CodeLlama checkpoints.

### PR DESCRIPTION
**Pull request by kpal002**
## What this fixes

Closes #46491.

`decoders.Strip(content=" ", left=1)` in `CodeLlamaTokenizer.__init__` was unconditionally removing one leading space on every `decode` call. For inputs that begin with real spaces (e.g. indented Python code), this silently corrupted the output — `decode(encode("    if x:\n"))` returned `"   if x:\n"` instead of `"    if x:\n"`.

The stripping was intended to remove the synthetic `▁` prefix that `Metaspace(prepend_scheme="first")` prepends before the first content token. The bug: after `Replace(▁ → " ")`, the synthetic prefix and real user-provided leading spaces are indistinguishable in the decoded string, so `Strip(left=1)` ate one real space whenever the input started with one or more spaces.

## Root cause

```python
# before (tokenization_code_llama.py line 163-164)
self._tokenizer.decoder = decoders.Sequence(
    [decoders.Replace("▁", " "), decoders.ByteFallback(), decoders.Fuse(), decoders.Strip(content=" ", left=1)]
)
```

## Fix

Remove `Strip` from the decoder sequence and override `_decode` to perform the strip conditionally, based on the first real (non-special) token:

- If the first real token consists **entirely of `▁` characters** (`▁`=29871, `▁▁`=259, `▁▁▁`=1678, …), it was produced by real user-provided leading spaces → **do not strip**.
- Otherwise, the leading space in the decoded text came from the synthetic Metaspace prefix → **strip one space**.

This handles 2, 3, 4, and N leading spaces, and is correct in the presence of a prepended BOS token (BOS is in `all_special_ids` and is skipped when looking for the first real token).

Note: a **single** leading space (`" hello"` vs `"hello"`) is inherently ambiguous — both encode to the same token IDs — and cannot round-trip. This was true before this fix and remains true; it is documented in the new test.

## Before / after

```
# transformers 5.10.x (before)
BUG: '  hello'           -> ' hello'
BUG: '   leading spaces' -> '  leading spaces'
BUG: '    indented_line' -> '   indented_line'

# after this fix
OK:  '  hello'           -> '  hello'
OK:  '   leading spaces' -> '   leading spaces'
OK:  '    indented_line' -> '    indented_line'
OK:  'hello world'       -> 'hello world'   # unaffected
```

## Tests

Added `test_leading_space_roundtrip` to `tests/models/code_llama/test_tokenization_code_llama.py` covering 2-, 3-, and 4-space-prefixed strings via `encode(s, add_special_tokens=False)` → `decode(…, skip_special_tokens=True)`.

Ran locally on this PR's branch (`pytest tests/models/code_llama/test_tokenization_code_llama.py`, Python 3.12.13, transformers 5.10.0.dev0 editable install):
```
60 passed, 8 skipped in 34.11s
```
including `CodeLlamaTokenizationTest::test_leading_space_roundtrip`. Also green on CI: https://circleci.com/gh/huggingface/transformers/2354143

## Disclosure

This fix was drafted with AI-agent assistance (diagnosis, patch, and test). I reviewed every changed line, confirmed the root cause by tracing the `super().__init__()` overwrite of `self._tokenizer` and the BPE merge behavior for multi-▁ tokens, and ran the test suite locally (output above) before opening this PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @arthurzucker @itazap

**Comment by qgallouedec** [comment]
closing as superseded, see https://github.com/huggingface/transformers/issues/47487#issuecomment-5143765657, thanks!
